generator/road_access_generator.cpp

```diff
@@ -0,0 +1,170 @@
+char constexpr kDelim[] = " \t\r\n";
+
+bool ParseRoadAccess(string const & roadAccessPath,
+                     map<uint64_t, uint32_t> const & osmIdToFeatureId,
+                     routing::RoadAccess & roadAccess)
+{
+  ifstream stream(roadAccessPath);
+  if (!stream)
+  {
+    LOG(LWARNING, ("Could not open", roadAccessPath));
+    return false;
+  }
+
+  string line;
+  for (uint32_t lineNo = 1;; ++lineNo)
+  {
+    if (!getline(stream, line))
+      break;
+
+    strings::SimpleTokenizer iter(line, kDelim);
+
+    if (!iter)
+    {
+      LOG(LWARNING, ("Error when parsing road access: empty line", lineNo));
+      return false;
+    }
+
+    string const s = *iter;
+    ++iter;
+
+    uint64_t osmId;
+    if (!iter || !strings::to_uint64(*iter, osmId))
+    {
+      LOG(LWARNING, ("Error when parsing road access: bad osm id at line", lineNo));
+      return false;
+    }
+
+    auto const it = osmIdToFeatureId.find(osmId);
+    if (it == osmIdToFeatureId.cend())
+    {
+      LOG(LWARNING, ("Error when parsing road access: unknown osm id", osmId, "at line", lineNo));
+      return false;
+    }
+
+    uint32_t const featureId = it->second;
+    roadAccess.GetPrivateRoads().emplace_back(featureId);
+  }
+
+  return true;
+}
+}  // namespace
+
+namespace routing
+{
+// RoadAccessWriter ------------------------------------------------------------
+void RoadAccessWriter::Open(string const & filePath)
+{
+  LOG(LINFO,
+      ("Saving information about barriers and road access classes in osm id terms to", filePath));
+  m_stream.open(filePath, ofstream::out);
+
+  if (!IsOpened())
+    LOG(LINFO, ("Cannot open file", filePath));
+}
+
+void RoadAccessWriter::Process(OsmElement const & elem, FeatureParams const & params)
+{
+  if (!IsOpened())
+  {
+    LOG(LWARNING, ("Tried to write to a closed barriers writer"));
+    return;
+  }
+
+  auto const & c = classif();
+
+  StringIL const forbiddenRoadTypes[] = {
+    {"hwtag", "private"}
+  };
+
+  for (auto const & f : forbiddenRoadTypes)
+  {
+    auto const t = c.GetTypeByPath(f);
+    if (params.IsTypeExist(t) && elem.type == OsmElement::EntityType::Way)
+      m_stream << kAccessPrivate << " " << elem.id << endl;
+  }
+
+  auto t = c.GetTypeByPath({"barrier", "gate"});
+  if (params.IsTypeExist(t))
+    m_stream << kBarrierGate << " " << elem.id << endl;
+}
+
+bool RoadAccessWriter::IsOpened() const { return m_stream.is_open() && !m_stream.fail(); }
+// RoadAccessCollector ----------------------------------------------------------
+RoadAccessCollector::RoadAccessCollector(string const & roadAccessPath,
+                                         string const & osmIdsToFeatureIdsPath)
+{
+  map<uint64_t, uint32_t> osmIdToFeatureId;
+  if (!ParseOsmIdToFeatureIdMapping(osmIdsToFeatureIdsPath, osmIdToFeatureId))
+  {
+    LOG(LWARNING, ("An error happened while parsing feature id to osm ids mapping from file:",
+                   osmIdsToFeatureIdsPath));
+    m_valid = false;
```

Comment:
Please, set m_valid to false in ctor initializer, and set to true only when both checks are passed.